### PR TITLE
Add FractionCSP3, LogS and HeteroAtomsCount

### PR DESCRIPTION
**Pull request by WhisperWind22**
FractionCSP3 and LogS

### CODE REVIEW

File: mordred/LogS.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    '[ch0]': -0.37570,
+    '[ch1]': -0.22435,
+    'F': -0.21728,
+    'Cl': -0.49721,
+    'Br': -0.57982,
+    'I': -0.51547
+}
+
+
+class LogS(Descriptor):
```
Comment (philopon) [inline mordred/LogS.py]: This class requires some tests.

File: mordred/LogS.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        return "LogS"
+
+    def __str__(self):
+        return self._type
+
+    def __init__(self, type="LogS"):
+        self._type = type
+        self._smart_logs_molecules = []
+        for key in smarts_logs:
+            self._smart_logs_molecules.append(Chem.MolFromSmarts(key))
```
Comment (philopon) [inline mordred/LogS.py]: _smart_log_molecules is not need as instance variable.
Please put toplevel to improve performance (please refer comment on L63).

File: mordred/LogS.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    def parameters(self):
+        return ()
+
+    def calculate(self, MW):
+        logS = 0.89823 - 0.10369 * math.sqrt(MW)
+        for index,key in enumerate(smarts_logs):
+            logS += len(self.mol.GetSubstructMatches(self._smart_logs_molecules[index])) * smarts_logs[key]
```
Comment (philopon) [inline mordred/LogS.py]: this procedure is little confusing.

In top level:

```.py
_smarts_logs_molecules = [
    (Chem.MolFromSmarts(smarts), log)
    for smarts, log in smarts_logs.items()
]
```

and, in calculate method:

```.py
for smarts, log in _smarts_logs_molecules:
    logS += len(self.mol.GetSubstructMatches(smarts)) * log
```

it helps understanding code and may bit fast.

File: mordred/LogS.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+        return logS
+
+    rtype = int
```
Comment (philopon) [inline mordred/LogS.py]: float?

**Comment by WhisperWind22** [comment]
Fixed
Thanks for review.

**Comment by WhisperWind22** [comment]
Hello!
Still need some changes?

**Comment by philopon** [comment]
I'm sorry for my late reply :cry: Your PR is almost perfect!

Could you tell me the source of test example of LogS (article, book, or other software etc.) ?

I'll merge your PR after CI is fixed.

**Comment by WhisperWind22** [comment]
I calculated through http://www.swissadme.ch/index.php

**Comment by philopon** [comment]
I read ESOL article [1] (it says original ESOL uses CLogP), SwissADME article [2] (it says ESOL in SwissADME uses XLogP3), and XLogP3 article [3] (it says XLogP3 uses 83 atom types and 6 extra features).

[1] https://doi.org/10.1021/ci034243x
[2] https://www.nature.com/articles/srep42717
[3] https://doi.org/10.1021/ci700257y

**Comment by WhisperWind22** [comment]
The formula for calculating LogS is taken from the silicos-it (filter-it) library:
http://silicos-it.be.s3-website-eu-west-1.amazonaws.com/software/filter-it/1.0.2/filter-it.html#logs

You could also check http://www.swissadme.ch - they by default use filter-it library. So the code that I implemented is based on the source code of the filter-it library. So the results of swissadme output have to be the same as in my code.

**Comment by philopon** [comment]
I got it :smile:

After some fixing, all values in [mordred test molecules](../blob/master/mordred/tests/references/structures.smi) are matched to values of filter-it program (25198cb9f12a5959e62e0a52a90c0ae90df37780).
Now, values of SwissADME is not reproduced. however, I think this implementation is suitable:

1. Filter-it is referred by SwissADME
2. From numerical example:

benzene
-0.22435 ([ch1]) * 6 - sqrt(78.114) + 0.89823 = -1.3643

In SwissADME, it seems that [ch0] * 6 are matched (it can reproduce when SMARTs matching to molecules that have explicit hydrogens in RDKit).

And, I think many LogS prediction models are reported. Thus, current name is too general.
Could I rename `LogS` to `FilterItLogS` (00544fb19062c5a15945bd95f499fb827204d64c)? I'll merge soon when you allow it.

**Comment by WhisperWind22** [comment]
Yes, of course.

**Comment by WhisperWind22** [comment]
I added 2 more descriptors. One for Plane of Best Fit (https://github.com/rdkit/rdkit/tree/master/Contrib/PBF, https://pubs.acs.org/doi/abs/10.1021/ci300293f), the other for calculating the ratio acyclic to cyclic SP3.

**Comment by philopon** [comment]
I am sorry to have kept you waiting.
FilterItLogS, FractionCSP3 and nHetero descriptors were merged into master.

your new commit a75c9aba62e4c4de3804e148d3454665154a42d5 seems good.
However, it is not merged in this PR, sorry...
Would you mind split PR SP3A2C and PBF?

Splitting PR per the kind of descriptor makes review easy.

Cheers,

**Pull request closed without merging**